### PR TITLE
[Feature] External Contributions

### PR DESCRIPTION
**Pull request by hvilloria**
## What does this PR do?

* request github API to get all projects from rootstrap members and process its pull requests filtering by pull request owners that belongs to rootstrap organization.
* refactors the sidebar since open source tab now uses it too.
* conditionals in sidebar are replaced by content_for

[Jira ticket](https://rootstrap.atlassian.net/browse/EM-112)

external_contributions

### CODE REVIEW

**Comment by hvilloria** [comment]
@hdamico @SandroDamilano sorry for the 50 files 🤦  a lot of them are just adding some space for `<%=end%> / do%>` tags

**Comment by santiagovidal** [comment]
UX comment: can we add a link to the Open Source sidebar that says "General metrics" and sends you to the page with the weekly repository visits? Otherwise, the only way to get back to that page is by clicking on the "Open Source" link at the top, which is not ideal.

File: app/jobs/external_contributions_processor_job.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+class ExternalContributionsProcessorJob < ApplicationJob
```
[the remark this answers is not in the document]
Comment (SandroDamilano) [reply app/jobs/external_contributions_processor_job.rb:1]: Should it be called [here](https://github.com/rootstrap/rs-code-review-metrics/pull/317/files?file-filters%5B%5D=.erb&file-filters%5B%5D=.json&file-filters%5B%5D=.rb&file-filters%5B%5D=.scss&file-filters%5B%5D=.sql#diff-9221cdf4e70e21a55f79894641fd706bR58)?

File: app/controllers/users/external_pull_requests_controller.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    def index
+      @external_pull_requests = ExternalPullRequest
+                                .where(created_at: from.weeks.ago..Time.zone.now)
+                                .group_by { |pull_request| pull_request.owner.login }
+    end
+
+    private
+
+    def from
+      params[:from].to_i || 4
```
[the remark this answers is not in the document]
Comment (hvilloria) [reply app/controllers/users/external_pull_requests_controller.rb:14]: i would like to move it into a service or something like that but since the query is pretty short, it goes well here in the controller, that's why i putted it there.. what do you think?

File: app/jobs/external_contributions_processor_job.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (hvilloria) [reply app/jobs/external_contributions_processor_job.rb:1]: from the schedule, it is a worker

File: app/services/builders/external_project.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+module Builders
+  class ExternalProject < BaseService
+    def initialize(repository_data)
+      @repository_data = repository_data
+    end
+
+    def call
+      ::ExternalProject.find_or_initialize_by(github_id: @repository_data[:id]) do |project|
```
[the remark this answers is not in the document]
Comment (hvilloria) [reply app/services/builders/external_project.rb:8]: yep, because we do not save it on cases where the projects returned by github API does not have any pull request for the user we are processing, so we just want to find/initialize to check later if is necessary persist it

File: app/services/builders/external_pull_request.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
+module Builders
+  class ExternalPullRequest < BaseService
```
[the remark this answers is not in the document]
Comment (hvilloria) [reply app/services/builders/external_pull_request.rb:2]: i wasn't sure when deciding that, i tested the classes around this "execution" but then i got into some redundance when i was going to test this, would you test the creation or query for the model ?

File: app/services/builders/external_pull_request.rb [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    def initialize(pull_request_data, external_project)
+      @pull_request_data = pull_request_data
+      @external_project = external_project
+    end
+
+    def call
+      ::ExternalPullRequest.find_or_create_by!(github_id: @pull_request_data[:id]) do |pull_request|
```
[the remark this answers is not in the document]
Comment (hvilloria) [reply app/services/builders/external_pull_request.rb:9]: it was kind of the intention, at the end of the execution we need to create it or doing nothing, since we don't even track the status of these pull requests i didn't want to be doing DB executions, i think the most important attribute here is pull_request's url, so that's why i didn't put to much effort keeping track of the pull request.. do you think we should update the rest information of the pr anyway?

File: app/jobs/external_contributions_processor_job.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (SandroDamilano) [reply app/jobs/external_contributions_processor_job.rb:1]: I know, but from the `schedule.yml` we're calling `ProcessExternalContributionsJob`, not `ExternalContributionsProcessorJob`

File: app/controllers/users/external_pull_requests_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (SandroDamilano) [reply app/controllers/users/external_pull_requests_controller.rb:14]: No, I definitely prefer having it here, the closer to the origin the better.
Now that I think about it, my doubt here is about the default number chosen itself. Why 4? It sounds like a good default for the frontend, not for the backend. The backend needs a default just to avoid breaking. And a 0 would be a perfect fit for that. If we receive a request without that param, I wouldn't think of answering with a logic response.
I know I'm being picky here, so I don't mind if you don't agree 😅

File: app/services/builders/external_pull_request.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (SandroDamilano) [reply app/services/builders/external_pull_request.rb:2]: Yeah, I know that feeling about redundancy haha
I tend to prefer that redundancy anyway, though. Because this class should be a standalone one and not necessarily be coupled with the other one.
As for this particular case, you could test the external pull request is created if not present and just returned if present. The related class that uses this one, `Processors::External:: PullRequests`, performs other checks and processes different payloads, so even though they are similar, they are not exactly the same tests.

File: app/services/builders/external_pull_request.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (SandroDamilano) [reply app/services/builders/external_pull_request.rb:9]: Not at all! If that was the intention, I'm perfectly fine with it. It makes perfect sense! 👍

File: app/services/github_client/repository.rb [hunk 6]
```diff
@@ -10,6 +10,14 @@ def code_owners
       find_in_locations
     end
 
+    def pull_requests
+      response = connection.get("repos/#{@project.full_name}/pulls") do |request|
```
[the remark this answers is not in the document]
Comment (hvilloria) [reply app/services/github_client/repository.rb:14]: thats great! im going to change it! 👍

File: app/services/github_client/repository.rb [hunk 7]
```diff
@@ -10,6 +10,14 @@ def code_owners
...
+        request['state'] = 'all'
+      end
+      response_body = response.body
+      response_body.empty? ? [] : JSON.parse(response_body, symbolize_names: true)
```
[the remark this answers is not in the document]
Comment (hvilloria) [reply app/services/github_client/repository.rb:18]: when there are no repos found, it has almost 0 probabilities 😅  but if it happens it will crash the execution

File: app/services/processors/external/contributions.rb [hunk 8]
```diff
@@ -0,0 +1,21 @@
+module Processors
+  module External
+    class Contributions < BaseService
+      def call
+        usernames.each do |username|
+          repositories_data = Processors::External::Repositories.call(username)
+          repositories_data.each do |repository_data|
+            project = Builders::ExternalProject.call(repository_data)
+            Processors::External::PullRequests.call(project, username)
```
[the remark this answers is not in the document]
Comment (hvilloria) [reply app/services/processors/external/contributions.rb:9]: im not sure if i understood, we could change the check of arrays in response by status to know what we should return, is that what you mean?

File: app/views/open_source/index.erb [hunk 9]
```diff
@@ -1,3 +1,15 @@
...
+<% content_for :open_source_sidebar do %>
+  <div class="btn-group btn-entity-group">
+    <div class='custom-list d-flex flex-row align-items-center'>
+      <div class='circle'></div>
+      <%= link_to external_pull_requests_path do %>
+        <span class='font-size-1'>External Contributions</span>
+      <% end %>
+    </div>
+  </div>
+<% end %>
```
[the remark this answers is not in the document]
Comment (hvilloria) [reply app/views/open_source/index.erb:11]: i thought something like put it in a partial, but i seemed weird to me, what do you think?

File: spec/services/processors/external/contributions_spec.rb [hunk 10]
```diff
@@ -0,0 +1,37 @@
+require 'rails_helper'
+
+RSpec.describe Processors::External::Contributions do
```
[the remark this answers is not in the document]
Comment (hvilloria) [reply spec/services/processors/external/contributions_spec.rb:3]: it could seems to be useless but since it is the "entry point" of the worker it is just the handler/caller of the other processors for every username we have in DB

File: app/controllers/users/external_pull_requests_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (hvilloria) [reply app/controllers/users/external_pull_requests_controller.rb:14]: no, it has a lot of sense to me, i will change that 👍

File: app/services/processors/external/contributions.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (SandroDamilano) [reply app/services/processors/external/contributions.rb:9]: Yeah, I'm sorry, I'll clarify on that.
A few weeks ago I added the `:raise_error` middleware to the requests made to Github, so if the response is not successful, an error will be raised.
We need to consider that any failed request will break the code execution, so we need to plan ahead and think of what we should do in such cases.
What I have opted to do is to let all processes related to that single request fail (because I could do nothing with the failed response) and catch the error when other processes that could finish correctly may be compromised. And track the error occurrence in Exception Hunter.
You can check an example [here](https://github.com/rootstrap/rs-code-review-metrics/blob/develop/app/services/processors/open_source_project_views_updater.rb), where I catch the error and let the process using that class continue with the execution normally.

File: app/views/open_source/index.erb (hunk 9 above)
[the remark this answers is not in the document]
Comment (SandroDamilano) [reply app/views/open_source/index.erb:11]: Yeah, I thought of a partial as well, as I've seen that in other places.
Why do you think it could be weird? I may be missing something 😅

File: spec/services/processors/external/contributions_spec.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (SandroDamilano) [reply spec/services/processors/external/contributions_spec.rb:3]: Yeah, I understand why you did it that way. But we are then checking _how_ and not _what_.
I think we could get rid of the mocks easily.
If we are checking that `Builders::ExternalProject` is called, we could assert that an external project for the user is created. And as for `Processors::External::PullRequests`, we could assert external pull requests have been created.
We don't need to go through all the scenarios those classes handle (like if those records existed before), we just need to make sure those records exist for all the users in the DB.
What do you think?

File: app/services/github_client/repository.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (SandroDamilano) [reply app/services/github_client/repository.rb:18]: But in that case, doesn't it return an empty list? I mean, the body itself is not empty, it's an empty list.

File: app/controllers/open_source_controller.rb [hunk 11]
```diff
@@ -1,4 +1,6 @@
 class OpenSourceController < ApplicationController
+  layout 'sidebar_metrics'
```
Comment (santib) [inline app/controllers/open_source_controller.rb:2]: why do we have 2 different layouts?

File: app/services/builders/external_pull_request.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (santib) [reply app/services/builders/external_pull_request.rb:2]: my criteria is something like:
- each class should be thoroughly tested with unit tests. this mean that unit tests should test a lot of different cases, including border cases
- integration tests should test only the most important flows and might check a reduced set of things, their focus is testing that some action is accomplished correctly

File: spec/factories/external_pull_requests.rb [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+#
+#  fk_rails_...  (external_project_id => external_projects.id)
+#  fk_rails_...  (owner_id => users.id)
+#
+
+FactoryBot.define do
+  factory :external_pull_request do
+    html_url { 'https://github.com/Codertocat/Hello-World/pull/2' }
+    title { "Pull Request-#{Faker::Number.number(digits: 1)}" }
+    github_id { Faker::Number.number(digits: 10) }
```
Comment (santib) [inline spec/factories/external_pull_requests.rb:30]: maybe we can use FactoryBot `sequence` for these?

File: spec/jobs/external_contibutions_processor_job_spec.rb [hunk 13]
```diff
@@ -0,0 +1,10 @@
+require 'rails_helper'
+require 'sidekiq/testing'
```
Comment (santib) [inline spec/jobs/external_contibutions_processor_job_spec.rb:2]: why is this needed? aren't we using the activejob API (not sidekiq API) ?

File: spec/jobs/external_contibutions_processor_job_spec.rb [hunk 14]
```diff
@@ -0,0 +1,10 @@
...
+
+RSpec.describe ExternalContributionsProcessorJob do
+  it 'executes external contributions processor' do
+    expect(Processors::External::Contributions).to receive(:call)
```
Comment (santib) [inline spec/jobs/external_contibutions_processor_job_spec.rb:6]: I'm ok with this test, but I see it as an integration test, so to add more value I'd have checked the happy path, that something was created or whatever

File: spec/services/processors/external/contributions_spec.rb [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+  let!(:user) { create(:user, login: 'hvilloria') }
+  context 'when there are repos for the given username' do
+    before do
+      stub_get_repos_from_user(user.login)
+      allow(Builders::ExternalProject).to receive(:call) {}
+      allow(Processors::External::PullRequests).to receive(:call) {}
```
Comment (santib) [inline spec/services/processors/external/contributions_spec.rb:9]: why is everything stubbed?

File: app/controllers/open_source_controller.rb (hunk 11 above)
Comment (hvilloria) [reply app/controllers/open_source_controller.rb:2]: because of the tab in navbar (Tech Blog) which does not use the sidebar

File: spec/services/processors/external/contributions_spec.rb (hunk 15 above)
Comment (hvilloria) [reply spec/services/processors/external/contributions_spec.rb]: it was because of the external requests, but i changed it by request stub so all the flow could get completed 👍

File: app/services/github_client/repository.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (hvilloria) [reply app/services/github_client/repository.rb:18]: i think i understand you point 🤔 , that is redundant.. let met change it a bit

**Review by SandroDamilano — APPROVED** [review]
Awesome work!!! 👏 🥇

File: app/services/builders/external_pull_request.rb [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+        pull_request.external_project_id = built_or_saved_project.id
+      end
+    end
+
+    private
+
+    def built_or_saved_project
+      return @external_project if @external_project.persisted?
+
+      @external_project.tap(&:save!)
```
[the remark this answers is not in the document]
Comment (hvilloria) [reply app/services/builders/external_pull_request.rb:23]: yes 😞

File: app/services/github_client/repository.rb [hunk 17]
```diff
@@ -15,7 +15,7 @@ def pull_requests
         request['state'] = 'all'
       end
       response_body = response.body
-      response_body.empty? ? [] : JSON.parse(response_body, symbolize_names: true)
+      JSON.parse(response_body, symbolize_names: true) unless response_body.empty?
```
Comment (SandroDamilano) [inline app/services/github_client/repository.rb:18]: I'm sorry, but I'll have to ask again: Is there any scenario where the response body will be empty?

File: app/assets/stylesheets/metrics/_interval-filter.scss [hunk 18]
```diff
@@ -11,3 +11,8 @@
     margin: 0 0.2rem;
   }
 }
+
+.period-input-external-pr{
```
Comment (santib) [inline app/assets/stylesheets/metrics/_interval-filter.scss:15]:
```suggestion
.period-input-external-pr {
```

File: app/services/builders/external_pull_request.rb (hunk 16 above)
Comment (santib) [inline app/services/builders/external_pull_request.rb:23]: why is that we don't save the record if it was already persisted? we don't want to update the attributes of the external PR?

File: app/services/rootstrap_member_repositories_discriminator.rb [hunk 19]
```diff
@@ -0,0 +1,15 @@
...
+    @repositories = repositories
+  end
+
+  def call
+    @repositories.reject { |repository| rs_members.include? repository[:owner][:login] }
+  end
+
+  private
+
+  def rs_members
```
Comment (santib) [inline app/services/rootstrap_member_repositories_discriminator.rb:12]: not sure if we want to make this repo "abstract" of RS or not. cc @santiagovidal

File: app/views/development_metrics/_departments_dropdown.erb [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+      <% if params[:department_name].blank? %>
+        <option selected value disabled>Choose a department</option>
+      <% else %>
+        <option value disabled>Choose a department</option>
+      <% end %>
+      <% all_departments_names.each do |name|%>
+        <%if params[:department_name] == name %>
+          <option selected value=<%= name %>> <%= name%> </option>
+        <% else %>
+          <option value=<%= name%>> <%= name%></option> %>
```
Comment (santib) [inline app/views/development_metrics/_departments_dropdown.erb:13]:
```suggestion
          <option value=<%= name %>> <%= name %></option> %>
```

File: app/views/development_metrics/_departments_dropdown.erb [hunk 21]
```diff
@@ -0,0 +1,18 @@
+<%= content_for :departments_dropdown do %>
+  <div class="list-department">
+    <select name="department_name" class="department-selection" form='nav-filter-form'>
+      <% if params[:department_name].blank? %>
+        <option selected value disabled>Choose a department</option>
+      <% else %>
+        <option value disabled>Choose a department</option>
+      <% end %>
+      <% all_departments_names.each do |name|%>
```
Comment (santib) [inline app/views/development_metrics/_departments_dropdown.erb:9]:
```suggestion
      <% all_departments_names.each do |name| %>
```

File: app/views/development_metrics/_departments_dropdown.erb [hunk 22]
```diff
@@ -0,0 +1,18 @@
...
+        <%if params[:department_name] == name %>
```
Comment (santib) [inline app/views/development_metrics/_departments_dropdown.erb:10]:
```suggestion
        <% if params[:department_name] == name %>
```

File: app/views/development_metrics/_development_metrics_sidebar.erb [hunk 23]
```diff
@@ -0,0 +1,3 @@
+<%= content_for :development_metrics_sidebar do %>
+  <%=  render 'development_metrics/sidebar'%>
```
Comment (santib) [inline app/views/development_metrics/_development_metrics_sidebar.erb:2]:
```suggestion
  <%= render 'development_metrics/sidebar'%>
```

File: app/views/development_metrics/projects.erb [hunk 24]
```diff
@@ -1,3 +1,6 @@
+<%= render 'development_metrics/projects_dropdown' %>
+<%=  render 'development_metrics/development_metrics_sidebar' %>
```
Comment (santib) [inline app/views/development_metrics/projects.erb:2]:
```suggestion
<%= render 'development_metrics/development_metrics_sidebar' %>
```

**Review by santib — APPROVED** [review]
Good job 👍

File: app/services/rootstrap_member_repositories_discriminator.rb (hunk 19 above)
Comment (santiagovidal) [reply app/services/rootstrap_member_repositories_discriminator.rb:12]: If possible, yes. In this case, we can name it `company_members` instead of `rs_members` for example.

File: app/services/builders/external_pull_request.rb (hunk 16 above)
Comment (hvilloria) [reply app/services/builders/external_pull_request.rb:23]: yes, we are not tracking almost anything about the external_project, in fact we only use the github_id just to get all the pull request and for pull requests we only want the html_url attribute, that's why i did it like that, im not sure if changes like full name or name will change things for us when processing.. besides that we want to avoid unnecessary queries to the DB because of the number of projects o pull request that a member could have
